{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f56f1c1e",
   "metadata": {},
   "source": [
    "Classes from BioRED corpus:\n",
    "- 'CellLine'\n",
    "- 'ChemicalEntity'\n",
    "- 'DiseaseOrPhenotypicFeature'\n",
    "- 'GeneOrGeneProduct'\n",
    "- 'OrganismTaxon'\n",
    "- 'SequenceVariant'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a06deb4f",
   "metadata": {},
   "source": [
    "Importing libraries:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "486ae23d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import nltk\n",
    "from nltk.tokenize import WordPunctTokenizer\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5ca201d",
   "metadata": {},
   "source": [
    "Converting files from .txt into  u neophodan oblik za treniranje:\n",
    "- text vectors are separated sentences,\n",
    "- in numeric vectors there are also those that consist only of zeros if they are not shorter than 3 tokens,\n",
    "- here the presence of NER is recorded only by ones, not by twos (2=1)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5de83cd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Training set\n",
    "\n",
    "address = './Dataset/Train.PubTator' # Dataset is a folder which should contain downloaded Train, Dev and Test datasets\n",
    "\n",
    "with open(address) as f:\n",
    "    \n",
    "    id_nums = list()\n",
    "    ts = list()\n",
    "    titles = list()\n",
    "    abstracts = list()\n",
    "    original_texts = list()\n",
    "    texts = list()\n",
    "    results = list()\n",
    "    original_results = list()\n",
    "    vectors = list()\n",
    "    corpuses = list()\n",
    "    sentences = list()\n",
    "    original_sentences = list()\n",
    "    \n",
    "    for x in f.read().split('\\n\\n'):\n",
    "        \n",
    "        try:\n",
    "            id_num = x.split('|t|')[0]\n",
    "            if id_num in ['17000021', '17006606', '9294109', '20086182', '18310445', '28398555',\n",
    "                         '30442153', '15184369', '16157158', '15198485', '17951029', '21666969',\n",
    "                          '19825989', '21496008', '16321363', '16506214', '18791947', '20801104',\n",
    "                          '28851297', '20651814', '17042910', '18340638', '19721134', '20683499']:\n",
    "                continue\n",
    "            else:\n",
    "                id_num = id_num\n",
    "            \n",
    "            title = x.split('|t|')[1].split('\\n')[0].rstrip()\n",
    "            if title.endswith('?'):\n",
    "                title = title[:-1] + '.'\n",
    "            elif title.endswith('.'):\n",
    "                title = title\n",
    "            else:\n",
    "                title = title + '.'\n",
    "                \n",
    "            abstract = x.split('|a|')[1].split('\\n')[0].rstrip()\n",
    "            if abstract.endswith('..'):\n",
    "                abstract = abstract[:-1]\n",
    "            elif abstract.endswith('.'):\n",
    "                abstract = abstract\n",
    "            else:\n",
    "                abstract = abstract + '.'\n",
    "\n",
    "            original_text = title + \" \" + abstract\n",
    "            \n",
    "            tags = list()\n",
    "            for y in x.split('\\n')[2:]:\n",
    "                pocetni = y.split('\\t')[1]\n",
    "                if pocetni.isdigit():\n",
    "                    tag = y.split('\\t')[4]\n",
    "                    tags.append(tag)\n",
    "            \n",
    "            for t in set(tags):\n",
    "                text = original_text\n",
    "                pomeraj = 0\n",
    "                for y in x.split('\\n')[2:]:\n",
    "                    pocetni = y.split('\\t')[1]\n",
    "                    if pocetni.isdigit():\n",
    "                        pocetni = int(pocetni) - pomeraj\n",
    "                        krajnji = int(y.split('\\t')[2]) - pomeraj\n",
    "                        NER, tag = y.split('\\t')[3:5]\n",
    "                        \n",
    "                        if tag == t:\n",
    "                            NER_len = len(WordPunctTokenizer().tokenize(NER))\n",
    "                            if NER_len == 1:\n",
    "                                a = 'xxxxx'\n",
    "                            else:\n",
    "                                a = 'xxxxx'+' yyyyy'*(NER_len-1)\n",
    "                        \n",
    "                            text = text[:pocetni] + a + text[krajnji:]\n",
    "                        \n",
    "                            pomeraj += krajnji - pocetni - len(a)\n",
    "                    else:\n",
    "                        continue\n",
    "                result = WordPunctTokenizer().tokenize(text)\n",
    "                original_result = WordPunctTokenizer().tokenize(original_text)\n",
    "                if len(result)==len(original_result):\n",
    "                    pass\n",
    "                else:\n",
    "                    print(id_num, 'NIJE OK', len(result), len(original_result))\n",
    "                \n",
    "                n = 0\n",
    "                for m in original_result:\n",
    "                    if m == '.' or m == ').':\n",
    "                        k = original_result.index(m, n)\n",
    "                        sentence = result[n:k+1]\n",
    "                        original_sentence = original_result[n:k+1]\n",
    "                        n = k + 1\n",
    "                        if len(sentence)==len(original_sentence):\n",
    "                            pass\n",
    "                        else:\n",
    "                            print(id_num, 'NIJE OK', sentence, original_sentence)\n",
    "                        vector = list()\n",
    "                        for i in sentence:\n",
    "                            if i=='xxxxx':\n",
    "                                i=1\n",
    "                            elif i=='yyyyy':\n",
    "                                i=1\n",
    "                            else:\n",
    "                                i=0\n",
    "                            vector.append(i)\n",
    "                        corpus = 'BioRED'\n",
    "                        if len(vector) >= 3:\n",
    "                            id_nums.append(id_num)\n",
    "                            ts.append(t)\n",
    "                            titles.append(title)\n",
    "                            abstracts.append(abstract)\n",
    "                            original_texts.append(original_text)\n",
    "                            texts.append(text)\n",
    "                            results.append(result)\n",
    "                            original_results.append(original_result)\n",
    "                            corpuses.append(corpus)\n",
    "                            sentences.append(sentence)\n",
    "                            original_sentences.append(original_sentence)\n",
    "                            vectors.append(vector)\n",
    "        except:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "07539805",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_train = pd.DataFrame(list(zip(corpuses, id_nums, ts, original_sentences, vectors)),\n",
    "               columns =['cor', 'ids', 'class', 'text', 'labels'])\n",
    "df_train = df_train.set_index('ids')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "52d6cfc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Development set\n",
    "\n",
    "address = './Dataset/Dev.PubTator'\n",
    "\n",
    "with open(address) as f:\n",
    "    \n",
    "    id_nums = list()\n",
    "    ts = list()\n",
    "    titles = list()\n",
    "    abstracts = list()\n",
    "    original_texts = list()\n",
    "    texts = list()\n",
    "    results = list()\n",
    "    original_results = list()\n",
    "    vectors = list()\n",
    "    corpuses = list()\n",
    "    sentences = list()\n",
    "    original_sentences = list()\n",
    "    \n",
    "    for x in f.read().split('\\n\\n'):\n",
    "        \n",
    "        try:\n",
    "            id_num = x.split('|t|')[0]\n",
    "            if id_num in ['16867246', '19880293', '27103577', '17600377', '25589620', '20621845']:\n",
    "                continue\n",
    "            else:\n",
    "                id_num = id_num\n",
    "            \n",
    "            title = x.split('|t|')[1].split('\\n')[0].rstrip()\n",
    "            if title.endswith('?'):\n",
    "                title = title[:-1] + '.'\n",
    "            elif title.endswith('.'):\n",
    "                title = title\n",
    "            else:\n",
    "                title = title + '.'\n",
    "                \n",
    "            abstract = x.split('|a|')[1].split('\\n')[0].rstrip()\n",
    "            if abstract.endswith('..'):\n",
    "                abstract = abstract[:-1]\n",
    "            elif abstract.endswith('.'):\n",
    "                abstract = abstract\n",
    "            else:\n",
    "                abstract = abstract + '.'\n",
    "\n",
    "            original_text = title + \" \" + abstract\n",
    "            \n",
    "            tags = list()\n",
    "            for y in x.split('\\n')[2:]:\n",
    "                pocetni = y.split('\\t')[1]\n",
    "                if pocetni.isdigit():\n",
    "                    tag = y.split('\\t')[4]\n",
    "                    tags.append(tag)\n",
    "            \n",
    "            for t in set(tags):\n",
    "                text = original_text\n",
    "                pomeraj = 0\n",
    "                for y in x.split('\\n')[2:]:\n",
    "                    pocetni = y.split('\\t')[1]\n",
    "                    if pocetni.isdigit():\n",
    "                        pocetni = int(pocetni) - pomeraj\n",
    "                        krajnji = int(y.split('\\t')[2]) - pomeraj\n",
    "                        NER, tag = y.split('\\t')[3:5]\n",
    "                        \n",
    "                        if tag == t:\n",
    "                            NER_len = len(WordPunctTokenizer().tokenize(NER))\n",
    "                            if NER_len == 1:\n",
    "                                a = 'xxxxx'\n",
    "                            else:\n",
    "                                a = 'xxxxx'+' yyyyy'*(NER_len-1)\n",
    "                        \n",
    "                            text = text[:pocetni] + a + text[krajnji:]\n",
    "                        \n",
    "                            pomeraj += krajnji - pocetni - len(a)\n",
    "                    else:\n",
    "                        continue\n",
    "                result = WordPunctTokenizer().tokenize(text)\n",
    "                original_result = WordPunctTokenizer().tokenize(original_text)\n",
    "                if len(result)==len(original_result):\n",
    "                    pass\n",
    "                else:\n",
    "                    print(id_num, 'NIJE OK', len(result), len(original_result))\n",
    "                \n",
    "                n = 0\n",
    "                for m in original_result:\n",
    "                    if m == '.' or m == ').':\n",
    "                        k = original_result.index(m, n)\n",
    "                        sentence = result[n:k+1]\n",
    "                        original_sentence = original_result[n:k+1]\n",
    "                        n = k + 1\n",
    "                        if len(sentence)==len(original_sentence):\n",
    "                            pass\n",
    "                        else:\n",
    "                            print(id_num, 'NIJE OK', sentence, original_sentence)\n",
    "                        vector = list()\n",
    "                        for i in sentence:\n",
    "                            if i=='xxxxx':\n",
    "                                i=1\n",
    "                            elif i=='yyyyy':\n",
    "                                i=1\n",
    "                            else:\n",
    "                                i=0\n",
    "                            vector.append(i)\n",
    "                        corpus = 'BioRED'\n",
    "                        if len(vector) >= 3:\n",
    "                            id_nums.append(id_num)\n",
    "                            ts.append(t)\n",
    "                            titles.append(title)\n",
    "                            abstracts.append(abstract)\n",
    "                            original_texts.append(original_text)\n",
    "                            texts.append(text)\n",
    "                            results.append(result)\n",
    "                            original_results.append(original_result)\n",
    "                            corpuses.append(corpus)\n",
    "                            sentences.append(sentence)\n",
    "                            original_sentences.append(original_sentence)\n",
    "                            vectors.append(vector)\n",
    "        except:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "259672b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_dev = pd.DataFrame(list(zip(corpuses, id_nums, ts, original_sentences, vectors)),\n",
    "               columns =['cor', 'ids', 'class', 'text', 'labels'])\n",
    "df_dev = df_dev.set_index('ids')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e9e9b6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test set\n",
    "\n",
    "address = './Dataset/Test.PubTator'\n",
    "\n",
    "with open(address) as f:\n",
    "    \n",
    "    id_nums = list()\n",
    "    ts = list()\n",
    "    titles = list()\n",
    "    abstracts = list()\n",
    "    original_texts = list()\n",
    "    texts = list()\n",
    "    results = list()\n",
    "    original_results = list()\n",
    "    vectors = list()\n",
    "    corpuses = list()\n",
    "    sentences = list()\n",
    "    original_sentences = list()\n",
    "    \n",
    "    for x in f.read().split('\\n\\n'):\n",
    "        \n",
    "        try:\n",
    "            id_num = x.split('|t|')[0]\n",
    "            if id_num in ['15485686', '15818664', '19276632', '28846666', '24442316', '18507837']:\n",
    "                continue\n",
    "            else:\n",
    "                id_num = id_num\n",
    "            \n",
    "            title = x.split('|t|')[1].split('\\n')[0].rstrip()\n",
    "            if title.endswith('?'):\n",
    "                title = title[:-1] + '.'\n",
    "            elif title.endswith('.'):\n",
    "                title = title\n",
    "            else:\n",
    "                title = title + '.'\n",
    "                \n",
    "            abstract = x.split('|a|')[1].split('\\n')[0].rstrip()\n",
    "            if abstract.endswith('..'):\n",
    "                abstract = abstract[:-1]\n",
    "            elif abstract.endswith('.'):\n",
    "                abstract = abstract\n",
    "            else:\n",
    "                abstract = abstract + '.'\n",
    "\n",
    "            original_text = title + \" \" + abstract\n",
    "            \n",
    "            tags = list()\n",
    "            for y in x.split('\\n')[2:]:\n",
    "                pocetni = y.split('\\t')[1]\n",
    "                if pocetni.isdigit():\n",
    "                    tag = y.split('\\t')[4]\n",
    "                    tags.append(tag)\n",
    "            \n",
    "            for t in set(tags):\n",
    "                text = original_text\n",
    "                pomeraj = 0\n",
    "                for y in x.split('\\n')[2:]:\n",
    "                    pocetni = y.split('\\t')[1]\n",
    "                    if pocetni.isdigit():\n",
    "                        pocetni = int(pocetni) - pomeraj\n",
    "                        krajnji = int(y.split('\\t')[2]) - pomeraj\n",
    "                        NER, tag = y.split('\\t')[3:5]\n",
    "                        \n",
    "                        if tag == t:\n",
    "                            NER_len = len(WordPunctTokenizer().tokenize(NER))\n",
    "                            if NER_len == 1:\n",
    "                                a = 'xxxxx'\n",
    "                            else:\n",
    "                                a = 'xxxxx'+' yyyyy'*(NER_len-1)\n",
    "                        \n",
    "                            text = text[:pocetni] + a + text[krajnji:]\n",
    "                        \n",
    "                            pomeraj += krajnji - pocetni - len(a)\n",
    "                    else:\n",
    "                        continue\n",
    "                result = WordPunctTokenizer().tokenize(text)\n",
    "                original_result = WordPunctTokenizer().tokenize(original_text)\n",
    "                if len(result)==len(original_result):\n",
    "                    pass\n",
    "                else:\n",
    "                    print(id_num, 'NIJE OK', len(result), len(original_result))\n",
    "                \n",
    "                n = 0\n",
    "                for m in original_result:\n",
    "                    if m == '.' or m == ').':\n",
    "                        k = original_result.index(m, n)\n",
    "                        sentence = result[n:k+1]\n",
    "                        original_sentence = original_result[n:k+1]\n",
    "                        n = k + 1\n",
    "                        if len(sentence)==len(original_sentence):\n",
    "                            pass\n",
    "                        else:\n",
    "                            print(id_num, 'NIJE OK', sentence, original_sentence)\n",
    "                        vector = list()\n",
    "                        for i in sentence:\n",
    "                            if i=='xxxxx':\n",
    "                                i=1\n",
    "                            elif i=='yyyyy':\n",
    "                                i=1\n",
    "                            else:\n",
    "                                i=0\n",
    "                            vector.append(i)\n",
    "                        corpus = 'BioRED'\n",
    "                        if len(vector) >= 3:\n",
    "                            id_nums.append(id_num)\n",
    "                            ts.append(t)\n",
    "                            titles.append(title)\n",
    "                            abstracts.append(abstract)\n",
    "                            original_texts.append(original_text)\n",
    "                            texts.append(text)\n",
    "                            results.append(result)\n",
    "                            original_results.append(original_result)\n",
    "                            corpuses.append(corpus)\n",
    "                            sentences.append(sentence)\n",
    "                            original_sentences.append(original_sentence)\n",
    "                            vectors.append(vector)\n",
    "        except:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0b92e668",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_test = pd.DataFrame(list(zip(corpuses, id_nums, ts, original_sentences, vectors)),\n",
    "               columns =['cor', 'ids', 'class', 'text', 'labels'])\n",
    "df_test = df_test.set_index('ids')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c22b2a16",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([df_train, df_dev, df_test])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ebb54e9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['class'] = df['class'].apply(lambda x: x.replace('CellLine', 'Cell Line'))\n",
    "df['class'] = df['class'].apply(lambda x: x.replace('ChemicalEntity', 'Chemical Entity'))\n",
    "df['class'] = df['class'].apply(lambda x: x.replace('DiseaseOrPhenotypicFeature', 'Disease Or Phenotypic Feature'))\n",
    "df['class'] = df['class'].apply(lambda x: x.replace('GeneOrGeneProduct', 'Gene Or Gene Product'))\n",
    "df['class'] = df['class'].apply(lambda x: x.replace('OrganismTaxon', 'Organism Taxon'))\n",
    "df['class'] = df['class'].apply(lambda x: x.replace('SequenceVariant', 'Sequence Variant'))\n",
    "df['class'] = df['class'].apply(lambda x: x.split(' '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "25d440b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['labels'] = df['labels'].apply(lambda x: np.array(x))\n",
    "df['labels'] = df['labels'].apply(lambda x: x.astype(np.int64))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f26a4800",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cor</th>\n",
       "      <th>class</th>\n",
       "      <th>text</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ids</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10491763</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[Hepatocyte, nuclear, factor, -, 6, :, associa...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10491763</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[The, transcription, factor, hepatocyte, nucle...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10491763</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[We, therefore, tested, the, hypothesis, that,...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10491763</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[We, cloned, the, coding, region, as, well, as...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10491763</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[We, then, examined, them, on, genomic, DNA, i...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24840785</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[The, results, showed, that, aconitine, stimul...</td>\n",
       "      <td>[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24840785</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[The, expression, analysis, of, Ca, (, 2, +), ...</td>\n",
       "      <td>[0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24840785</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[The, expression, analysis, of, apoptosis, -, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24840785</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[Furthermore, ,, increased, phosphorylation, o...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24840785</th>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[Hence, ,, our, results, suggest, that, aconit...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29900 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             cor                class  \\\n",
       "ids                                     \n",
       "10491763  BioRED  [Sequence, Variant]   \n",
       "10491763  BioRED  [Sequence, Variant]   \n",
       "10491763  BioRED  [Sequence, Variant]   \n",
       "10491763  BioRED  [Sequence, Variant]   \n",
       "10491763  BioRED  [Sequence, Variant]   \n",
       "...          ...                  ...   \n",
       "24840785  BioRED   [Chemical, Entity]   \n",
       "24840785  BioRED   [Chemical, Entity]   \n",
       "24840785  BioRED   [Chemical, Entity]   \n",
       "24840785  BioRED   [Chemical, Entity]   \n",
       "24840785  BioRED   [Chemical, Entity]   \n",
       "\n",
       "                                                       text  \\\n",
       "ids                                                           \n",
       "10491763  [Hepatocyte, nuclear, factor, -, 6, :, associa...   \n",
       "10491763  [The, transcription, factor, hepatocyte, nucle...   \n",
       "10491763  [We, therefore, tested, the, hypothesis, that,...   \n",
       "10491763  [We, cloned, the, coding, region, as, well, as...   \n",
       "10491763  [We, then, examined, them, on, genomic, DNA, i...   \n",
       "...                                                     ...   \n",
       "24840785  [The, results, showed, that, aconitine, stimul...   \n",
       "24840785  [The, expression, analysis, of, Ca, (, 2, +), ...   \n",
       "24840785  [The, expression, analysis, of, apoptosis, -, ...   \n",
       "24840785  [Furthermore, ,, increased, phosphorylation, o...   \n",
       "24840785  [Hence, ,, our, results, suggest, that, aconit...   \n",
       "\n",
       "                                                     labels  \n",
       "ids                                                          \n",
       "10491763  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "10491763  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "10491763  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "10491763  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "10491763  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "...                                                     ...  \n",
       "24840785                  [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]  \n",
       "24840785  [0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, ...  \n",
       "24840785  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "24840785  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "24840785  [0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, ...  \n",
       "\n",
       "[29900 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d11b5152",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>cor</th>\n",
       "      <th>class</th>\n",
       "      <th>text</th>\n",
       "      <th>labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10491763</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[Hepatocyte, nuclear, factor, -, 6, :, associa...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10491763</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[The, transcription, factor, hepatocyte, nucle...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10491763</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[We, therefore, tested, the, hypothesis, that,...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10491763</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[We, cloned, the, coding, region, as, well, as...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10491763</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Sequence, Variant]</td>\n",
       "      <td>[We, then, examined, them, on, genomic, DNA, i...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29895</th>\n",
       "      <td>24840785</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[The, results, showed, that, aconitine, stimul...</td>\n",
       "      <td>[0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29896</th>\n",
       "      <td>24840785</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[The, expression, analysis, of, Ca, (, 2, +), ...</td>\n",
       "      <td>[0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29897</th>\n",
       "      <td>24840785</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[The, expression, analysis, of, apoptosis, -, ...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29898</th>\n",
       "      <td>24840785</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[Furthermore, ,, increased, phosphorylation, o...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29899</th>\n",
       "      <td>24840785</td>\n",
       "      <td>BioRED</td>\n",
       "      <td>[Chemical, Entity]</td>\n",
       "      <td>[Hence, ,, our, results, suggest, that, aconit...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29900 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ids     cor                class  \\\n",
       "0      10491763  BioRED  [Sequence, Variant]   \n",
       "1      10491763  BioRED  [Sequence, Variant]   \n",
       "2      10491763  BioRED  [Sequence, Variant]   \n",
       "3      10491763  BioRED  [Sequence, Variant]   \n",
       "4      10491763  BioRED  [Sequence, Variant]   \n",
       "...         ...     ...                  ...   \n",
       "29895  24840785  BioRED   [Chemical, Entity]   \n",
       "29896  24840785  BioRED   [Chemical, Entity]   \n",
       "29897  24840785  BioRED   [Chemical, Entity]   \n",
       "29898  24840785  BioRED   [Chemical, Entity]   \n",
       "29899  24840785  BioRED   [Chemical, Entity]   \n",
       "\n",
       "                                                    text  \\\n",
       "0      [Hepatocyte, nuclear, factor, -, 6, :, associa...   \n",
       "1      [The, transcription, factor, hepatocyte, nucle...   \n",
       "2      [We, therefore, tested, the, hypothesis, that,...   \n",
       "3      [We, cloned, the, coding, region, as, well, as...   \n",
       "4      [We, then, examined, them, on, genomic, DNA, i...   \n",
       "...                                                  ...   \n",
       "29895  [The, results, showed, that, aconitine, stimul...   \n",
       "29896  [The, expression, analysis, of, Ca, (, 2, +), ...   \n",
       "29897  [The, expression, analysis, of, apoptosis, -, ...   \n",
       "29898  [Furthermore, ,, increased, phosphorylation, o...   \n",
       "29899  [Hence, ,, our, results, suggest, that, aconit...   \n",
       "\n",
       "                                                  labels  \n",
       "0      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "1      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "2      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "3      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "4      [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "...                                                  ...  \n",
       "29895                  [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0]  \n",
       "29896  [0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, ...  \n",
       "29897  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "29898  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...  \n",
       "29899  [0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 0, ...  \n",
       "\n",
       "[29900 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.reset_index()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "74d209a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# using protocol 4 for dumping a .pkl file\n",
    "\n",
    "filename = './BioRED_protokol4.pkl'\n",
    "outfile = open(filename,'wb')\n",
    "pickle.dump(df,outfile, protocol=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "dcc7ba14",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_pickle('./BioRED_ivi.pkl')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
